cg/meta/workflow/microsalt/quality_controller/utils.py

```diff
@@ -0,0 +1,142 @@
+
+
+def has_valid_average_coverage(metrics: SampleMetrics) -> bool:
+    coverage: float | None = metrics.microsalt_samtools_stats.average_coverage
+    return is_valid_average_coverage(coverage) if coverage else False
+
+
+def has_valid_10x_coverage(metrics: SampleMetrics) -> bool:
+    coverage_10x: float | None = metrics.microsalt_samtools_stats.coverage_10x
+    return is_valid_10x_coverage(coverage_10x) if coverage_10x else False
+
+
+def get_negative_control_result(results: list[SampleQualityResult]) -> SampleQualityResult:
+    for result in results:
+        if result.is_control:
+            return result
+    raise ValueError("No negative control found")
+
+
+def negative_control_pass_qc(results: list[SampleQualityResult]) -> bool:
+    negative_control_result: SampleQualityResult = get_negative_control_result(results)
+    return negative_control_result.passes_qc
+
+
+def get_results_passing_qc(results: list[SampleQualityResult]) -> list[SampleQualityResult]:
+    return [result for result in results if result.passes_qc]
+
+
+def get_non_urgent_results(results: list[SampleQualityResult]) -> list[SampleQualityResult]:
+    return [result for result in results if not is_urgent_result(result)]
+
+
+def get_urgent_results(results: list[SampleQualityResult]) -> list[SampleQualityResult]:
+    return [result for result in results if is_urgent_result(result)]
+
+
+def is_urgent_result(result: SampleQualityResult) -> bool:
+    return result.application_tag == MicrosaltAppTags.MWRNXTR003
+
+
+def urgent_samples_pass_qc(results: list[SampleQualityResult]) -> bool:
+    urgent_results: list[SampleQualityResult] = get_urgent_results(results)
+    return all(result.passes_qc for result in urgent_results)
+
+
+def non_urgent_samples_pass_qc(results: list[SampleQualityResult]) -> bool:
+    non_urgent_samples: list[SampleQualityResult] = get_non_urgent_results(results)
+    passing_qc: list[SampleQualityResult] = get_results_passing_qc(non_urgent_samples)
+
+    if not non_urgent_samples:
+        return True
+
+    fraction_passing_qc: float = len(passing_qc) / len(non_urgent_samples)
+    return fraction_passing_qc >= MicrosaltQC.QC_PERCENT_THRESHOLD_MWX
+
+
+def is_sample_negative_control(sample: Sample) -> bool:
+    return sample.control == ControlEnum.negative
+
+
+def get_application_tag(sample: Sample) -> str:
+    return sample.application_version.application.tag
+
+
+def get_sample_target_reads(sample: Sample) -> int:
+    return sample.application_version.application.target_reads
+
+
+def get_report_path(metrics_file_path: Path) -> Path:
+    return metrics_file_path.parent.joinpath(QUALITY_REPORT_FILE_NAME)
+
+
+def quality_control_case(sample_results: list[SampleQualityResult]) -> CaseQualityResult:
+    control_pass_qc: bool = negative_control_pass_qc(sample_results)
+    urgent_pass_qc: bool = urgent_samples_pass_qc(sample_results)
+    non_urgent_pass_qc: bool = non_urgent_samples_pass_qc(sample_results)
```

Comment:
I like this, we are going to transition into only having one apptag, and have the priority on the case be the type of analysis instead. Makes it easier to switch it around in these functions. Neat! ✨